### PR TITLE
Issue715 transformation cache

### PR DESCRIPTION
**Pull request by otbe**

### CODE REVIEW

**Comment by uweschaefer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/CacheKey.java  1
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uwe/factcast/pullRequest?prid=5234417&bid=17159456)

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/TransformationConfiguration.java [hunk 1]
```diff
@@ -41,6 +44,16 @@ public TransformationStore transformationStore(@NonNull JdbcTemplate jdbcTemplat
         return new InMemTransformationStoreImpl();
     }
 
+    @Bean
+    public TransformationCache transformationCache(@NonNull JdbcTemplate jdbcTemplate,
+            @NonNull PgConfigurationProperties props, @NonNull SpringLiquibase unused) {
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/TransformationConfiguration.java]: needs to be @Autowired(required=false) for liquibase

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/CacheKey.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package org.factcast.store.pgsql.registry.transformation.cache;
+
+import java.util.UUID;
+
+import org.factcast.core.Fact;
+
+import lombok.NonNull;
+
+public class CacheKey {
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/CacheKey.java]: better remove public

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/InMemTransformationCache.java [hunk 3]
```diff
@@ -31,22 +33,23 @@
...
+        String key = CacheKey.of(eventId, version, transformationChainId);
+
         return Optional.ofNullable(cache.get(key));
     }
 
     @Override
-    public void compact(DateTime thresholdDate) {
-        // TODO Auto-generated method stub
+    public void compact(@NonNull DateTime thresholdDate) {
+        cache.clear();
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/InMemTransformationCache.java]: harsh :laughing:

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class PgTransformationCache implements TransformationCache {
+    private final JdbcTemplate jdbcTemplate;
+
+    // note that SpringLiquibase needs to be injected in order to make sure it
+    // is initialized before we're intentionally not using @DependsOn here,
+    // as a change of the beanname within liquibase would break our code
+    private final SpringLiquibase liquibase;
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java]: please create constructor and @autowire(required=false) liquibase

File: factcast-store-pgsql/src/main/resources/db/changelog/factcast/issue715/transformationCache.sql [hunk 5]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE transformationcache(
+	cache_key 		varchar(2048) PRIMARY KEY,
+	header 			text NOT NULL,
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/main/resources/db/changelog/factcast/issue715/transformationCache.sql]: why not jsonb ? overhead?

File: factcast-store-pgsql/src/main/resources/db/changelog/factcast/issue715/transformationCache.sql [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+	payload 		text NOT NULL,
+	last_access 		TIMESTAMP WITH TIME ZONE DEFAULT now() NOT NULL
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/main/resources/db/changelog/factcast/issue715/transformationCache.sql]: i'd add a comment: intentionally not indexed (as i almost forgot why :D)

File: factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCacheTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    @Test
+    void testLastAccessUpdateAfterFind() {
+        Fact fact = Fact.builder()
+                .ns("ns")
+                .type("type")
+                .id(UUID.randomUUID())
+                .version(1)
+                .build("{}");
+        String chainId = "1-2-3";
+
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/test/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCacheTest.java]: someone loves blank lines :older_man:

File: factcast-store-pgsql/src/test/resources/test_schema.sql [hunk 8]
```diff
@@ -79,10 +80,15 @@ CREATE TABLE transformationstore(
...
-    UNIQUE(id)
+    transformation 	text
 );
 
 CREATE INDEX IF NOT EXISTS idx_transformationstore on transformationstore(ns,type);
 
- 
+ CREATE TABLE IF NOT EXISTS transformationcache(
+	cache_key 		varchar(2048) PRIMARY KEY,
+	header 			text NOT NULL,
```
Comment (uweschaefer) [inline factcast-store-pgsql/src/test/resources/test_schema.sql]: jsonb? see above

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/TransformationConfiguration.java (hunk 1 above)
Comment (otbe) [reply factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/TransformationConfiguration.java]: Changed it in both places

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java (hunk 4 above)
Comment (otbe) [reply factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java]: Mhh....`PgTransformationCache` is no bean/Component so I think theres no need for @autowired !?  🤔

**Comment by uweschaefer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java  2
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/CacheKey.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uwe/factcast/pullRequest?prid=5234417&bid=17159456)

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java (hunk 4 above)
Comment (uweschaefer) [reply factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java]: if it is not, then why is SpringLiquibase a dependency, anyway?

File: factcast-store-pgsql/src/main/resources/db/changelog/factcast/issue715/transformationCache.sql (hunk 5 above)
Comment (uweschaefer) [reply factcast-store-pgsql/src/main/resources/db/changelog/factcast/issue715/transformationCache.sql]: fine

**Comment by uweschaefer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/CacheKey.java  1
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uwe/factcast/pullRequest?prid=5234417&bid=17159456)

**Comment by uweschaefer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/PgTransformationCache.java  2
- factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/registry/transformation/cache/CacheKey.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/uwe/factcast/pullRequest?prid=5234417&bid=17159456)

**Pull request merged**